File: src/test/java/com/portfolio/saga/domain/SagaOrchestratorTest.java
package com.portfolio.saga.domain;

import com.portfolio.saga.application.InMemorySagaLog;
import org.junit.jupiter.api.Test;

import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

class SagaOrchestratorTest {

    @Test
    void shouldCompleteSagaWhenAllStepsSucceed() {
        InMemorySagaLog log = new InMemorySagaLog();
        Saga saga = new Saga(
                List.of(
                        new TestStep("step-1", false),
                        new TestStep("step-2", false),
                        new TestStep("step-3", false)
                ),
                log, "order-1"
        );

        SagaOrchestrator orchestrator = new SagaOrchestrator();
        SagaStatus result = orchestrator.execute(saga);

        assertEquals(SagaStatus.COMPLETED, result);
        assertEquals(SagaStatus.COMPLETED, saga.getStatus());
    }

    @Test
    void shouldExecuteStepsInOrder() {
        InMemorySagaLog log = new InMemorySagaLog();
        Saga saga = new Saga(
                List.of(
                        new TestStep("first", false),
                        new TestStep("second", false),
                        new TestStep("third", false)
                ),
                log, "order-1"
        );

        SagaOrchestrator orchestrator = new SagaOrchestrator();
        orchestrator.execute(saga);

        List<LogEntry> entries = log.entries(saga.getId());
        assertTrue(entries.stream().anyMatch(e -> e.message().contains("first succeeded")));
        assertTrue(entries.stream().anyMatch(e -> e.message().contains("second succeeded")));
        assertTrue(entries.stream().anyMatch(e -> e.message().contains("third succeeded")));
    }

    @Test
    void shouldRecordLogEntries() {
        InMemorySagaLog log = new InMemorySagaLog();
        Saga saga = new Saga(
                List.of(new TestStep("step-1", false)),
                log, "order-1"
        );

        SagaOrchestrator orchestrator = new SagaOrchestrator();
        orchestrator.execute(saga);

        List<LogEntry> entries = log.entries(saga.getId());
        assertFalse(entries.isEmpty());
    }

    private static class TestStep implements SagaStep<String, String> {
        private final String name;
        private final boolean shouldFail;

        TestStep(String name, boolean shouldFail) {
            this.name = name;
            this.shouldFail = shouldFail;
        }

        @Override
        public String getName() { return name; }

        @Override
        public String execute(String input) {
            if (shouldFail) throw new RuntimeException(name + " failed");
            return name + " executed";
        }

        @Override
        public void compensate(String input) {
        }
    }
}

File: src/test/kotlin/com/portfolio/saga/domain/SagaOrchestratorTest.kt
package com.portfolio.saga.domain

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertSame
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Test

class SagaOrchestratorTest {
    private val store = InMemorySagaStore()
    private val steps = OrderStep.entries.map(::IdempotentTestStep)
    private val orchestrator = SagaOrchestrator(store, steps)

    @Test
    fun `completes every operation exactly once`() {
        val result = orchestrator.start("order-complete")

        assertEquals(SagaStatus.COMPLETED, result.status)
        assertEquals(3, result.nextStep)
        steps.forEach { assertEquals(setOf("order-complete"), it.activeOrders) }
    }

    @Test
    fun `failure at every step reaches a consistent compensated state`() {
        OrderStep.entries.forEach { failedStep ->
            val orderId = "order-fail-${failedStep.position}"
            val result = orchestrator.start(orderId, FailurePlan(failStep = failedStep))

            assertEquals(SagaStatus.COMPENSATED, result.status)
            steps.take(failedStep.position).forEach { assertEquals(true, orderId in it.compensatedOrders) }
            steps.drop(failedStep.position).forEach { assertEquals(false, orderId in it.activeOrders) }
        }
    }

    @Test
    fun `compensation failure is FAILED until recovery succeeds`() {
        val failed = orchestrator.start(
            "order-compensation-failure",
            FailurePlan(
                failStep = OrderStep.CREATE_SHIPMENT,
                failCompensationStep = OrderStep.AUTHORIZE_PAYMENT,
            ),
        )

        assertEquals(SagaStatus.FAILED, failed.status)
        assertEquals(OrderStep.AUTHORIZE_PAYMENT.position, failed.compensationStep)

        val recovered = orchestrator.resume(failed.id)
        assertEquals(SagaStatus.COMPENSATED, recovered.status)
        assertEquals(emptySet<String>(), steps[0].activeOrders)
        assertEquals(emptySet<String>(), steps[1].activeOrders)
    }

    @Test
    fun `crash after side effect resumes idempotently from durable index`() {
        assertThrows(SimulatedCrashException::class.java) {
            orchestrator.start(
                "order-crash",
                FailurePlan(crashAfterStep = OrderStep.AUTHORIZE_PAYMENT),
            )
        }
        val interrupted = requireNotNull(store.getByOrderId("order-crash"))
        assertEquals(SagaStatus.RUNNING, interrupted.status)
        assertEquals(1, interrupted.nextStep)

        val recovered = orchestrator.resume(interrupted.id)
        assertEquals(SagaStatus.COMPLETED, recovered.status)
        assertEquals(2, steps[1].executionAttempts)
        assertEquals(setOf("order-crash"), steps[1].activeOrders)
    }

    @Test
    fun `order id is the idempotency boundary`() {
        val first = orchestrator.start("same-order")
        val second = orchestrator.start("same-order")

        assertEquals(first.id, second.id)
        assertSame(store.get(first.id), store.get(second.id))
        steps.forEach { assertEquals(1, it.executionAttempts) }
    }
}
